client/src/AppBundle/Controller/AbstractController.php

```diff
@@ -2,202 +2,44 @@
-            }
-        }
-
-        return $report;
-    }
-
-    /**
-     * @param int   $reportId
-     * @param array $groups
-     *
-     * @throws DisplayableException if report doesn't have specified section
-     * @throws \Symfony\Component\HttpKernel\Exception\NotFoundHttpException if report is submitted
-     *
-     * @return Report
-     */
-    protected function getReportIfNotSubmitted($reportId, array $groups = [])
-    {
-        $report = $this->getReport($reportId, $groups);
-
-        $sectionId = $this->getSectionId();
-        if ($sectionId && !$report->hasSection($sectionId)) {
-            throw new DisplayableException('Section not accessible with this report type.');
-        }
-
-        if ($report->getSubmitted()) {
-            throw new ReportSubmittedException();
-        }
-
-        return $report;
-    }
-
-    /**
-     * @param int   $ndrId
-     * @param array $groups
-     *
-     * @return Ndr
-     */
-    public function getNdr($ndrId, array $groups)
-    {
-        $groups[] = 'ndr';
-        $groups[] = 'ndr-client';
-        $groups[] = 'client';
-        $groups = array_unique($groups);
-
-        return $this->getRestClient()->get("ndr/{$ndrId}", 'Ndr\Ndr', $groups);
-    }
-
-    /**
-     * @param int $reportId
-     *
-     * @throws \Symfony\Component\HttpKernel\Exception\NotFoundHttpException if report is submitted
-     *
-     * @return Ndr
-     *
-     */
-    protected function getNdrIfNotSubmitted($reportId, array $groups = [])
+    protected function stepRedirector()
     {
-        $report = $this->getNdr($reportId, $groups);
-        if ($report->getSubmitted()) {
-            throw new ReportSubmittedException();
-        }
-
-        return $report;
+        /** @var StepRedirector */
+        return $this->get(StepRedirector::class);
     }
 
     /**
-     * @return MailFactory
+     * @param string $description
+     * @param int $statusCode
+     * @return Response
      */
-    protected function getMailFactory()
+    protected function renderError(string $description, $statusCode = 500)
     {
-        /** @var MailFactory */
-        $mailFactory = $this->get('AppBundle\Service\Mailer\MailFactory');
-        return $mailFactory;
-    }
+        $text = $this->renderView('TwigBundle:Exception:template.html.twig', [
+            'message' => 'Application error',
+            'description' => $description
+        ]);
 
-    /**
-     * @return MailSender
-     */
-    protected function getMailSender()
-    {
-        /** @var MailSender */
-        $mailSender = $this->get('AppBundle\Service\Mailer\MailSender');
-        return $mailSender;
+        return new Response($text, $statusCode);
     }
 
     /**
-     * @return Router
+     * @return object
      */
-    private function getRouter()
+    public function getRouter()
     {
         /** @var Router */
-        $router = $this->get('router');
-        return $router;
+        return $this->get('router');
```

Comment:
So where ever `getRouter` was used on its own I've injected the class instead -- but I've kept the method because of the `routeExists()` function we also have in the `AbstractController` class. 

The reason being `routeExists()` is used across 7 different classes all doing something similar to this:
```
$backLinkRoute = 'contacts_' . $request->get('from');
$backLink = $this->routeExists($backLinkRoute) ? $this->generateUrl($backLinkRoute, ['reportId'=>$reportId]) : '';
```

If we removed the `routeExists` function I feel like we'd be creating more code than necessary across the 7 classes by doing
```
$backLinkRoute = 'contacts_' . $request->get('from');
$backLinkTrue = $this->router->getRouteCollection()->get($backLinkRoute) ? true : false;
$backLink = $backLinkTrue ? $this->generateUrl($backLinkRoute, ['reportId' => $reportId]);
```

Only other option if we really did not want the `routeExists` function inside `AbstractController` is to create the function in all the 7 different classes but that's a lot of duplication. 

Happy to hear opinions :)

Reply:
^ Ignore that entire message 🤦🏻 I've just seen the other comment about testing route exists from Symfony 4